Analytics/Integrations/Kahuna/SEGKahunaIntegration.m

```diff
@@ -38,42 +103,400 @@ - (void)validate {
+        }
+        @catch (NSException *exception) {
+          // Do nothing.
+        }
+      }
+    }
+    
+    // Track the attributes if we have any items in it.
+    if (attributes.count > 0) {
+      [KahunaAnalytics setUserAttributes:attributes];
+    }
+  }
+  @catch (NSException *exception) {
+    NSLog (@"Kahuna-Segment Exception : %@", exception.description);
   }
-  [KahunaAnalytics setUserAttributes:traits];
 }
 
 - (void)track:(NSString *)event properties:(NSDictionary *)properties options:(NSDictionary *)options {
-  NSNumber *revenue = [self.class extractRevenue:properties];
-  if (revenue == nil){
-    [KahunaAnalytics trackEvent:event];
+  @try {
+    NSNumber *revenue = [self.class extractRevenue:properties];
+    NSNumber *quantity = nil;
+    for (NSString *key in properties) {
+      if ([key caseInsensitiveCompare:@"quantity"] == NSOrderedSame) {
+        id value = properties[key];
+        if ([value isKindOfClass:[NSString class]]) {
+          quantity = [NSNumber numberWithLong:[value longLongValue]];
+        } else if ([value isKindOfClass:[NSNumber class]]) {
+          quantity = value;
+        }
+        
+        break;
+      }
+    }
+    
+    // Get the count and value from quantity and revenue.
+    long value = (long) ([revenue doubleValue] * 100);
+    long count = [quantity longValue];
+    
+    if (count + value > 0) {
+      [KahunaAnalytics trackEvent:event withCount:count andValue:value];
+    } else {
+      [KahunaAnalytics trackEvent:event];
+    }
+    
+    NSMutableDictionary *attributes = [[NSMutableDictionary alloc] init];
+    NSMutableDictionary *lowerCaseKeyProperties = [[NSMutableDictionary alloc] init];
+    
+    // Lower case all the keys and copy over the properties into a new dictionary.
+    for (NSString *eachKey in properties) {
+      [lowerCaseKeyProperties setValue:properties[eachKey] forKey:[eachKey lowercaseString]];
+    }
+    
+    if ([event caseInsensitiveCompare:KAHUNA_VIEWED_PRODUCT_CATEGORY] == NSOrderedSame) {
+      [self addViewedProductCategoryElements:&attributes fromProperties:lowerCaseKeyProperties];
+    } else if ([event caseInsensitiveCompare:KAHUNA_VIEWED_PRODUCT] == NSOrderedSame) {
+      [self addViewedProductElements:&attributes fromProperties:lowerCaseKeyProperties];
+    } else if ([event caseInsensitiveCompare:KAHUNA_ADDED_PRODUCT] == NSOrderedSame) {
+      [self addAddedProductElements:&attributes fromProperties:lowerCaseKeyProperties];
+    } else if ([event caseInsensitiveCompare:KAHUNA_COMPLETED_ORDER] == NSOrderedSame) {
+      [self addCompletedOrderElements:&attributes fromProperties:lowerCaseKeyProperties];
+    }
+    
+    // If we have collected any attributes, then we will call the setUserAttributes API
+    if  (attributes.count > 0) {
+      [KahunaAnalytics setUserAttributes:attributes];
+    }
+  }
+  @catch (NSException *exception) {
+    NSLog (@"Kahuna-Segment Exception : %@", exception.description);
```

Comment:
All the Try-Catch are put in place as a safety mechanism for cases where the input dictionary keys or values change to un-anticipated types in the future or get modified to other object types that do not accept the called methods being made on them. Since this code once deployed cannot be pulled back with a new SDK release, this is purely added for defensive purposes.